lightning/src/ln/channelmanager.rs

```diff
@@ -2670,17 +2709,33 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 	}
 
 	fn internal_revoke_and_ack(&self, their_node_id: &PublicKey, msg: &msgs::RevokeAndACK) -> Result<(), MsgHandleErrInternal> {
-		let (pending_forwards, mut pending_failures, short_channel_id) = {
+		let (commitment_update, pending_forwards, mut pending_failures, closing_signed, monitor_update, htlcs_to_fail) = {
 			let mut channel_state_lock = self.channel_state.lock().unwrap();
 			let channel_state = &mut *channel_state_lock;
 			match channel_state.by_id.entry(msg.channel_id) {
 				hash_map::Entry::Occupied(mut chan) => {
 					if chan.get().get_their_node_id() != *their_node_id {
 						return Err(MsgHandleErrInternal::send_err_msg_no_close("Got a message for a channel from the wrong node!".to_owned(), msg.channel_id));
 					}
+					try_chan_entry!(self, chan.get_mut().revoke_and_ack(&msg, &self.fee_estimator, &self.logger), channel_state, chan)
+				},
+				hash_map::Entry::Vacant(_) => return Err(MsgHandleErrInternal::send_err_msg_no_close("Failed to find corresponding channel".to_owned(), msg.channel_id))
+			}
+		};
+
+		// Failing back holding cell HTLCs requires acquiring the channel state
```

Comment:
I don't think we can drop the channel_state lock between getting the return value from revoke_and_ack and handling it - otherwise we introduce race issues where we may eg drop the lock, have a different thread go forward HTLCs into the channel, and then take the lock and try to submit monitor updates etc. That would result in messages getting sent to the peer out of order or monitor updates happening out of order.

Reply:
That makes sense. To be clear, I'm interpreting this as: [these lines](https://github.com/rust-bitcoin/rust-lightning/blob/master/lightning/src/ln/channelmanager.rs#L2709) should also be moved to within the first lock (in addition to moving the holding-cell-htlc-failings to within the first lock).

Reply:
I think basically everything which was under the lock previously needs to remain under the lock, we just need to also be able to process the fail-backs even if we want to return early.